lib/dco.js

```diff
@@ -50,36 +52,101 @@ module.exports = async function (commits, isRequiredFor, prURL) {
     if (signoffs.length === 1) {
       // commit contains one signoff
       const sig = signoffs[0]
-      if (!(authors.includes(sig.name.toLowerCase())) || !(emails.includes(sig.email.toLowerCase()))) {
+      // if the signoff is an explicit match or an individual remediation, check if signoff name and email match the commit
+      // if the signoff is from a third-party remediation and passed the tests in getSignoffs, it is valid
+      if ((['explicit', 'individualRemediation'].includes(sig.type)) &&
+        (!(authors.includes(sig.name.toLowerCase())) ||
+        !(emails.includes(sig.email.toLowerCase())))) {
         commitInfo.message = `Expected "${commit.author.name} <${commit.author.email}>", but got "${sig.name} <${sig.email}>".`
         failed.push(commitInfo)
       }
     } else {
       // commit contains multiple signoffs
+      // first, find explicit signoffs which match the author exactly
       const valid = signoffs.filter(
-        signoff => authors.includes(signoff.name.toLowerCase()) && emails.includes(signoff.email.toLowerCase())
+        signoff => authors.includes(signoff.name.toLowerCase()) && emails.includes(signoff.email.toLowerCase()) && (signoff.type === 'explicit')
       )
+      // next, find individual remediations which match the author exactly
+      const validIndividualRemediation = signoffs.filter(
+        signoff => authors.includes(signoff.name.toLowerCase()) && emails.includes(signoff.email.toLowerCase()) && (signoff.type === 'individualRemediation')
+      )
+      // if the signoff is from a third-party remediation and passed the tests in getSignoffs, it is valid
+      const validThirdPartyRemediation = signoffs.filter(
+        signoff => signoff.type === 'thirdPartyRemediation')
 
-      if (valid.length === 0) {
+      // fail if no signoffs of any type match author's name and email
+      if ((valid.length === 0) && (validIndividualRemediation.length === 0) && (validThirdPartyRemediation.length === 0)) {
         const got = signoffs.map(sig => `"${sig.name} <${sig.email}>"`).join(', ')
         commitInfo.message = `Can not find "${commit.author.name} <${commit.author.email}>", in [${got}].`
         failed.push(commitInfo)
       }
     } // end if
   } // end for
-  return failed
+  if (failed) {
+    return failed.sort((a, b) => a.author.localeCompare(b.author))
+  } else {
+    return []
+  }
 }
 
-function getSignoffs (commit) {
-  const regex = /^Signed-off-by: (.*) <(.*)>$/img
+function getSignoffs (comparisonCommit, comparisonSha, allCommits, allowRemediationCommits) {
```

Comment:
I find the naming `comparison` a bit confusing - this is referring to the single commit that's currently being processed correct? esp when we get into [this area](https://github.com/probot/dco/pull/147/files#diff-ffe63918db292082603fd2f4cc677c991da9be21a8591c6503d4504efad0b822R115-R118) it gets pretty hard to parse.

Reply:
Yes, that's correct and it's a good point. It is confusing. `currentCommit` and `currentSha` might be better? As in, this is the current commit we're processing in an attempt to find valid remediation text in all other commits.